decimal/decimal.nim

```diff
@@ -6,3 +6,93 @@
 # at your option. This file may not be copied, modified, or distributed except according to those terms.
 
 import decimal_lowlevel
+
+type 
+    DecimalType = ptr mpd_t
+
+const
+    DEFAULT_PREC = MPD_RDIGITS * 2
+    DEFAULT_EMAX = when (sizeof(int) == 8): 999999999999999999 else: 425000000
+    DEFAULT_EMIN = when (sizeof(int) == 8): -999999999999999999 else: -425000000
+
+var CTX: mpd_context_t
+var CTX_ADDR = addr CTX
+mpd_defaultcontext(CTX_ADDR)
```

Comment:
There should be an `addQuitProc` statement that destroys/releases the context (or maybe a `defer`)

I don't find any related proc though :/. If you find one, note that you might have to add {.noconv.} pragma to that release proc otherwise addQuitProc will complain.

The main limitation of that approach of hiding the context is that for multi-threading you need one context per thread, but it's fine to start with single-threaded support and redirect people who need multi-threading to the low-level API.

Alternatively there is a `getContext` proc that checks if the current thread as a context already allocated, if yes retrieves it if no create one.

For that CTX should be a {.threadvar.}, you can see an example of this here: https://github.com/status-im/nim-eth-keys/blob/b09f5ee894183d8e49c02b0d9c21c617a7d1ce6b/eth_keys/libsecp256k1.nim#L50-L119

Reply:
Regarding the hiding of the context, I couldn't come up with a way to pass a context when doing for example `a + b`. Is there any way of doing that neatly?

Reply:
in 1990 it might have been ok to create a single-threaded api, but it feels a bit off nowadays - except perhaps that nim multi-threading is too broken to use for now ;) `addQuitProc` is pretty useless in general though - if the process is terminating, there's no need to release the context really as the OS will take care of it, and if it's aborting it won't run. 

Destroying contexts is needed mostly if you're doing a proper API that allows you to control the lifetime of the library or use it in multiple parts of your application independently and without leaks (ie plugins or modules that don't know about each other).

threadvar comes with its own problems including the point / time of initialization, and no (easy) way to coordinate release - what happens if you want to pass an decimal between threads?

are mpc contexts re-entrant/thread safe? many contexts are (because they contain only read-only stuff), in that case you don't need on per thread - I'd check the upstream docs here.

Reply:
Multiple threads reading from the same context is supported (https://www.bytereef.org/mpdecimal/doc/libmpdec/functions.html#thread-safety)

The context looks to be read-only: see [specification](http://speleotrove.com/decimal/decarith.html) and the [context part](http://speleotrove.com/decimal/daconcep.html)

This means that we can keep the currnet implementation without jumping through hoops.